# RFM Analizi İle Müşteri Segmentasyonu

## Görev 1: Veriyi Anlama ve Hazırlama

### Adım 1: Online Retail II excelindeki 2010-2011 verisini okuyunuz. Oluşturduğunuz dataframe'in kopyasını oluşturunuz.

In [1]:
import pandas as pd
import datetime as dt

In [2]:
df_ = pd.read_excel(r"C:\Users\myboz\OneDrive\Masaüstü\mybozkir\1. Projects\1. Miuul Data Analyst Path\6. CRM Analytics\datasets\online_retail_II.xlsx", sheet_name = 'Year 2010-2011')
df = df_.copy()

### Adım 2: Veri setinin betimsel istatistiklerini inceleyiniz.

In [3]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [5]:
df.shape

(541910, 8)

In [6]:
df.describe().T

count          mean          std       min       25%  \
Quantity     541910.0      9.552234   218.080957 -80995.00      1.00   
Price        541910.0      4.611138    96.759765 -11062.06      1.25   
Customer ID  406830.0  15287.684160  1713.603074  12346.00  13953.00   

                  50%       75%      max  
Quantity         3.00     10.00  80995.0  
Price            2.08      4.13  38970.0  
Customer ID  15152.00  16791.00  18287.0

### Adım 3: Veri setinde eksik gözlem var mı? Varsa hangi değişkende kaç tane eksik gözlem vardır?

In [8]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

Veri seti içerisinde, Description kolonunda 1454 adet, Customer ID kolonunda ise 135080 adet eksik değer vardır.

### Adım 4: Eksik gözlemleri veri setinden çıkartınız. Çıkarma işleminde 'inplace=True' parametresini kullanınız.

In [9]:
df.dropna(inplace=True)

In [10]:
df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
dtype: int64

### Adım 5: Eşsiz ürün sayısı kaçtır?

In [12]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

In [13]:
df['Description'].nunique()

3896

### Adım 6: Hangi üründen kaçar tane vardır?

In [14]:
df['Description'].value_counts().to_frame().head()

Description
WHITE HANGING HEART T-LIGHT HOLDER         2070
REGENCY CAKESTAND 3 TIER                   1905
JUMBO BAG RED RETROSPOT                    1662
ASSORTED COLOUR BIRD ORNAMENT              1418
PARTY BUNTING                              1416

### Adım 7: En çok sipariş edilen 5 ürünü çoktan aza doğru sıralayınız.

In [15]:
df['Description'].value_counts().sort_values(ascending=False).to_frame().head()

Description
WHITE HANGING HEART T-LIGHT HOLDER         2070
REGENCY CAKESTAND 3 TIER                   1905
JUMBO BAG RED RETROSPOT                    1662
ASSORTED COLOUR BIRD ORNAMENT              1418
PARTY BUNTING                              1416

### Adım 8: Faturalardaki 'C' iptal edilen işlemleri göstermektedir. İptal edilen işlemleri veri setinden çıkartınız.

In [19]:
df[~df['Invoice'].str.contains('C', na=False)].head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

In [20]:
df = df[~df['Invoice'].str.contains('C', na=False)]

### Adım 9: Fatura başına elde edilen toplam kazancı ifade eden 'TotalPrice' adında bir değişken oluşturunuz.

In [21]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

In [23]:
df['Total_Price'] = df['Quantity'] * df['Price']

In [24]:
df.Invoice.nunique()

18536

In [25]:
df.groupby('Invoice').agg({
    'Total_Price':'sum'
})

Total_Price
Invoice             
536365        139.12
536366         22.20
536367        278.73
536368         70.05
536369         17.85
...              ...
581583        124.60
581584        140.64
581585        329.05
581586        339.20
581587        267.45

[18536 rows x 1 columns]

## Görev 2: RFM Metriklerinin Hesaplanması

### Adım 1: Recency, Frequency ve Monetary tanımlarını yapınız.

In [31]:
today_date = dt.datetime(2011, 12, 11)

In [32]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  Total_Price  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom        15.30  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom        20.34  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom        22.00  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom        20.34  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom        20.34

### Adım 2: Müşteri özelinde Recency, Frequency, Monetary metriklerini groupby, agg ve lambda ile hesaplayınız.

### Adım 3: Hesapladığınız metrikleri rfm isimli bir değişkene atayınız.

In [33]:
rfm = df.groupby('Customer ID').agg({
    'InvoiceDate': lambda invoice_date: (today_date - invoice_date.max()).days,
    'Invoice': lambda invoice: invoice.nunique(),
    'Price': lambda price: price.sum()
})

In [34]:
rfm.head()

InvoiceDate  Invoice   Price
Customer ID                              
12346.0              326        1    1.04
12347.0                3        7  481.21
12348.0               76        4  178.71
12349.0               19        1  605.10
12350.0              311        1   65.30

### Adım 4: Oluşturduğunuz metriklerin isimlerini recency, frequency ve monetary olarak değiştiriniz.

In [35]:
rfm.columns = ['recency', 'frequency', 'monetary']

In [36]:
rfm.head()

recency  frequency  monetary
Customer ID                              
12346.0          326          1      1.04
12347.0            3          7    481.21
12348.0           76          4    178.71
12349.0           19          1    605.10
12350.0          311          1     65.30

## Görev 3: RFM Skorlarının Oluşturulması ve Tek Bir Değişkene Çevrilmesi

### Adım 1: Recency, Frequency ve Monetary metriklerini qcut yardımı ile 1-5 arasında skorlara çeviriniz.

### Adım 2: Bu skorları recency_score, frequency_score ve monetary-score olarak kaydediniz.

In [38]:
rfm['recency_score'] = pd.qcut(rfm['recency'], 5, labels=[5, 4, 3, 2, 1])

In [42]:
rfm['frequency_score'] = pd.qcut(rfm['frequency'].rank(method='first'), 5, labels=[1, 2, 3, 4, 5])

In [43]:
rfm['monetary_score'] = pd.qcut(rfm['monetary'], 5, labels=[1, 2, 3, 4, 5])

### Adım 3: recency_score ve frequency_score'u tek bir değişken olarak ifade ediniz ve RF_SCORE olarak kaydediniz.

In [44]:
rfm.head()

recency  frequency  monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1      1.04             1               1   
12347.0            3          7    481.21             5               5   
12348.0           76          4    178.71             2               4   
12349.0           19          1    605.10             4               1   
12350.0          311          1     65.30             1               1   

            monetary_score  
Customer ID                 
12346.0                  1  
12347.0                  5  
12348.0                  4  
12349.0                  5  
12350.0                  2

In [49]:
rfm['rfm_score'] = (rfm['recency_score'].astype('str') + rfm['frequency_score'].astype('str'))

In [50]:
rfm.head()

recency  frequency  monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1      1.04             1               1   
12347.0            3          7    481.21             5               5   
12348.0           76          4    178.71             2               4   
12349.0           19          1    605.10             4               1   
12350.0          311          1     65.30             1               1   

            monetary_score rf_score rfm_score  
Customer ID                                    
12346.0                  1       11        11  
12347.0                  5       55        55  
12348.0                  4       24        24  
12349.0                  5       41        41  
12350.0                  2       11        11

In [51]:
rfm = rfm.drop('rf_score', axis=1)

In [52]:
rfm.head()

recency  frequency  monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1      1.04             1               1   
12347.0            3          7    481.21             5               5   
12348.0           76          4    178.71             2               4   
12349.0           19          1    605.10             4               1   
12350.0          311          1     65.30             1               1   

            monetary_score rfm_score  
Customer ID                           
12346.0                  1        11  
12347.0                  5        55  
12348.0                  4        24  
12349.0                  5        41  
12350.0                  2        11

## Görev 4: RF Skorunun Segment Olarak Tanımlanması

### Adım 1: Oluşturulan RF skorları için segment tanımlamalarını yapınız.

### Adım 2: Regex ile skorları segmentlere çeviriniz.

In [47]:
seg_map = {
    r'[1-2][1-2]':'hibernating',
    r'[1-2][3-4]':'at_risk',
    r'[1-2]5':'cant_loose',
    r'3[1-2]':'about_to_sleep',
    r'33':'need_attention',
    r'[3-4][4-5]':'loyal_customers',
    r'41':'promising',
    r'51':'new_customers',
    r'[4-5][2-3]':'potential_loyalists',
    r'5[4-5]':'champions',
}

In [53]:
rfm['segment'] = rfm['rfm_score'].replace(seg_map, regex=True)

In [54]:
rfm.head()

recency  frequency  monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1      1.04             1               1   
12347.0            3          7    481.21             5               5   
12348.0           76          4    178.71             2               4   
12349.0           19          1    605.10             4               1   
12350.0          311          1     65.30             1               1   

            monetary_score rfm_score      segment  
Customer ID                                        
12346.0                  1        11  hibernating  
12347.0                  5        55    champions  
12348.0                  4        24      at_risk  
12349.0                  5        41    promising  
12350.0                  2        11  hibernating

## Görev 5 Aksiyon Zamanı!

### Adım 1: Önemli gördüğünüz 3 segmenti seçiniz. Bu üç segmenti hem aksiyon kararları açısından hem de segmentlerin yapısı açısından (RFM değerleri) yorumlayınız.

In [61]:
rfm[rfm['segment'].isin(['need_attention', 'promising', 'potential_loyalists'])][['segment', 'recency', 'frequency', 'monetary']].groupby('segment').agg(['mean', 'count'])

recency       frequency          monetary      
                          mean count      mean count        mean count
segment                                                               
need_attention       52.427807   187  2.326203   187  216.671979   187
potential_loyalists  17.398760   484  2.010331   484  140.578740   484
promising            23.421053    95  1.000000    95   67.584000    95

### Adım 2: "Loyal Customers" sınıfına ait customer ID'leri seçerek excel çıktısı alınız.

In [57]:
rfm[rfm['segment'] == 'loyal_customers'].to_csv('loyal_customers.csv')